pkg/util/fileIndexer.go

```diff
@@ -0,0 +1,165 @@
+package util
+
+import (
+	"encoding/json"
+	"github.com/golang/glog"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+const fileIndexDirectory = ".odo"
+const fileIndexName = "odo-file-index.json"
+
+type FileData struct {
+	Size             int64
+	LastModifiedDate time.Time
+}
+
+// read tries to read the odo index file from the given location and returns the data from the file
+// if no such file is present, it means the folder hasn't been walked and thus returns a empty list
+func read(filePath string) (map[string]FileData, error) {
+	// read operation
+	fileReadMap := make(map[string]FileData)
+	if _, err := os.Stat(filePath); os.IsNotExist(err) {
+		return map[string]FileData{}, nil
+	}
+
+	jsonFileRead, err := os.Open(filePath)
+	// if we os.Open returns an error then handle it
+	if err != nil {
+		return nil, err
+	}
+	defer jsonFileRead.Close()
+
+	byteValue, _ := ioutil.ReadAll(jsonFileRead)
+	err = json.Unmarshal(byteValue, &fileReadMap)
+	if err != nil {
+		return nil, err
+	}
+	return fileReadMap, nil
+}
+
+// resolveFilePath resolves the filepath of the odo index file in the .odo folder
+func resolveFilePath(directory string) (string, error) {
+	directoryFi, err := os.Stat(filepath.Join(directory))
+	if err != nil {
+		return "", err
+	}
+
+	switch mode := directoryFi.Mode(); {
+	case mode.IsDir():
+		// do directory stuff
+		return filepath.Join(directory, fileIndexDirectory, fileIndexName), nil
+	case mode.IsRegular():
+		// do file stuff
+		// for binary files
+		return filepath.Join(filepath.Dir(directory), fileIndexDirectory, fileIndexName), nil
+	}
+
+	return directory, nil
+}
+
+// Run walks the given directory and finds the files which have changed and which were deleted/renamed
+// it reads the odo index file from the .odo folder
+// if no such file is present, it means it's the first time the folder is being walked and thus returns a empty list
+// after the walk, it stores the list of walked files with some information in a odo index file in the .odo folder
+func Run(directory string, ignoreRules []string) (filesChanged []string, filesDeleted []string, err error) {
+	resolvedPath, err := resolveFilePath(directory)
+	if err != nil {
+		return filesChanged, filesDeleted, err
+	}
+
+	// read the odo index file
+	readData, err := read(resolvedPath)
+	if err != nil {
+		return filesChanged, filesDeleted, err
+	}
+
+	filesMap := make(map[string]FileData)
+	walk := func(fn string, fi os.FileInfo, err error) error {
+		if fi.IsDir() {
+
+			// if folder is the root folder, don't add it
+			if fi.Name() == filepath.Base(directory) && len(strings.Split(fi.Name(), "/")) == 1 {
```

Comment:
What about Windows?

Reply:
This should work on windows as `/` is used a separator there too and we are using golang libraries

Reply:
@mik-dass doesn't matter.. it won't change how path's are done, in windows it'll show either `C:/` or `C:\` for the path not `/home/user/foobar`

Reply:
I am not using the root folder. I am getting the base folder of the source code and Go will handle it e.g /home/user/nodejs or C://users/nodejs

Reply:
I will have a look into this on a windows machine

Reply:
@cdrage Fixed